### PR TITLE
Painless: ensure type "UnmodifiableMap" for params

### LINKED ISSUE #10561: [BUG] Flaky Runtime Exception while executing Painless script during Update call
**Describe the bug**

While performing updates through stored [Painless] scripts, we error out with `java.lang.UnsupportedOperationException -- java.util.Collections$UnmodifiableMap.put(Collections.java:1457) ~[?:?]` only so often.

**To Reproduce**

<details>

<summary>Steps to reproduce the behavior:</summary>

Checkout [`e4a1841`](https://github.com/opensearch-project/OpenSearch/commit/e4a1841e23dadd3351216383fdc16bf56a0b783d) from `opensearch:main`.

```sh
# run server
$ ./gradlew run

...

# Create Index
PUT sample-idx
{
  "mappings": {
    "properties": {
      "name": {
        "type": "text"
      },
      "rank": {
        "type": "integer"
      }
    }
  }
}

{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "sample-idx"
}

# Index a Doc
POST sample-idx/_doc/1
{
  "name": "Rohit",
  "rank": 1
}

{
  "_index": "sample-idx",
  "_id": "1",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "_seq_no": 0,
  "_primary_term": 1
}

# Store a sample script
POST _scripts/sample_script
{
  "script": {
    "lang": "painless",
    "source": """
    params.rank = 2
    """
  }
}

{
  "acknowledged": true
}

# Perform Update through Script
POST sample-idx/_update/1
{
  "script": {
    "id": "sample_script",
    "params": {
      "name": "r1walz"
    }
  }
}

# Now, it succeeds every now-and-then ...

{
  "_index": "sample-idx",
  "_id": "1",
  "_version": 2,
  "result": "updated",
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "_seq_no": 1,
  "_primary_term": 1
}

# ... but, fails more often than not

{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "failed to execute script"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "failed to execute script",
    "caused_by": {
      "type": "script_exception",
      "reason": "runtime error",
      "script_stack": [
        "java.base/java.util.Collections$UnmodifiableMap.put(Collections.java:1457)",
        "params.rank = 2\n    ",
        "      ^---- HERE"
      ],
      "script": "sample_script",
      "lang": "painless",
      "position": {
        "offset": 11,
        "start": 5,
        "end": 25
      },
      "caused_by": {
        "type": "unsupported_operation_exception",
        "reason": null
      }
    }
  },
  "status": 400
}

# Corresponding Stacktrace

java.lang.IllegalArgumentException: failed to execute script
        at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:342) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.UpdateHelper.prepareUpdateScriptRequest(UpdateHelper.java:279) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.UpdateHelper.prepare(UpdateHelper.java:109) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.UpdateHelper.prepare(UpdateHelper.java:90) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:242) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:235) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:90) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.support.single.instance.TransportInstanceSingleOperationAction$ShardTransportHandler$1.doRun(TransportInstanceSingleOperationAction.java:321) [opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:908) [opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-3.0.0.jar:3.0.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.opensearch.script.ScriptException: runtime error
        at org.opensearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:109) ~[?:?]
        at org.opensearch.painless.PainlessScript$Script.execute(sample_script:1) ~[?:?]
        at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:339) ~[opensearch-3.0.0.jar:3.0.0]
        ... 12 more
Caused by: java.lang.UnsupportedOperationException
        at java.util.Collections$UnmodifiableMap.put(Collections.java:1457) ~[?:?]
        at org.opensearch.painless.PainlessScript$Script.execute(sample_script:12) ~[?:?]
        at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:339) ~[opensearch-3.0.0.jar:3.0.0]
        ... 12 more

```

</details>

**Expected behavior**

Painless should be execute as per expectation without encountering any bugs. In this case though, it seems throwing exception is the desired behaviour as `params` Map is immutable once created and script seems to modify it. Either way, we should throw a better cause for `unsupported_operation_exception` instead for posting `null`.

**Plugins**

Standard.

**Screenshots**

https://github.com/opensearch-project/OpenSearch/assets/31043830/562a7fa6-3231-47b0-a3fd-cb70e78d650c

**Host/Environment (please complete the following information):**

- EC2 m5.2xlarge latest default ubuntu linux AMI

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by r1walz**
### Description

See comment here: https://github.com/opensearch-project/OpenSearch/issues/10561#issuecomment-2138270628

### Related Issues

Resolves #10561

### Check List
- [ ] ~New functionality includes testing.~
  - [ ] ~All tests pass~
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by r1walz** [comment]
It's a user facing bug so removed `skip-changelog` label. I'll add change log once a round of `gradle-check` finishes.

**Review by reta — APPROVED** [review]

**Comment by r1walz** [comment]
Updated `CHANGELOG.md` due to conflict.

NOTE: change log of `Pass parent filter to inner hit query` was updated to #13903 from #13770. Hence, the conflict.

**Review by gaobinlong — COMMENTED** [review]
Could we add a yml test case to verify we can always get the expected behavior?

**Comment by r1walz** [comment]
Hi. How can we specify number of nodes in the yml test? Also, how to send request to a specific node? I tried looking into the `rest-api-spec` folder and closest feature is `node_selector` but it only allows `version` and `attribute` as params. Can we even setup attribute through setup section? I'm lost here. Should we add integ test instead?

**Comment by reta** [comment]
> Hi. How can we specify number of nodes in the yml test?

So the YAML client tests by default run with one node (if I am not mistaken) but the `qa/*` submodules create different configuration (nodes / versions etc). I think the only criteria you need is to have (under `lang-painless/rest-api-specs` tests):

```
  - skip:
      version: "- 2.14.99"
      reason: "The params are unmodifiable since 2.15.0"

```

File: modules/lang-painless/src/test/java/org/opensearch/painless/WhenThingsGoWrongTests.java [hunk 1]
```diff
@@ -354,6 +354,9 @@ public void testInvalidAssignment() {
         assertEquals(iae.getMessage(), "invalid assignment: cannot assign a value to addition operation [+]");
         iae = expectScriptThrows(IllegalArgumentException.class, () -> exec("Double.x() = 1;"));
         assertEquals(iae.getMessage(), "invalid assignment: cannot assign a value to method call [x/0]");
+
+        expectScriptThrows(UnsupportedOperationException.class, () -> exec("params['modifyingParamsMap'] = 2;"));
```
Comment (dblock) [inline modules/lang-painless/src/test/java/org/opensearch/painless/WhenThingsGoWrongTests.java:358]: Nit: can we adopt the same pattern of `iae = ...` as above or change the above to 1-liners?

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
@r1walz Can you please backport manually? Appreciate it.

**Pull request merged**